url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/ZMQ.java

```diff
@@ -0,0 +1,814 @@
+    {
+        return Proxy.proxy(insocket_, outsocket_, null);
+    }
+
+    /**
+     * Polling on items. This has very poor performance.
+     * Try to use zmq_poll with selector
+     * CAUTION: This could be affected by jdk epoll bug
+     *
+     * @param items_
+     * @param timeout_
+     * @return number of events
+     */
+    public static int zmq_poll(PollItem[] items_, long timeout_) {
+        return zmq_poll(items_, items_.length, timeout_);
+    }
+
+    /**
+     * Polling on items. This has very poor performance.
+     * Try to use zmq_poll with selector
+     * CAUTION: This could be affected by jdk epoll bug
+     *
+     * @param items_
+     * @param timeout_
+     * @return number of events
+     */
+    public static int zmq_poll(PollItem[] items_, int count, long timeout_) {
+        Selector selector = null;
+        try {
+            selector = PollSelector.open();
+        } catch (IOException e) {
+            throw new ZError.IOException(e);
+        }
+
+        int ret = zmq_poll (selector, items_, count, timeout_);
+
+        //  Do not close selector
+        
+        return ret;
+    }
+
+    /**
+     * Polling on items with given selector
+     * CAUTION: This could be affected by jdk epoll bug
+     *
+     * @param selector Open and reuse this selector and do not forget to close when it is not used.
+     * @param items_
+     * @param timeout_
+     * @return number of events
+     */
+    public static int zmq_poll(Selector selector, PollItem[] items_, long timeout_)
+    {
+        return zmq_poll(selector, items_, items_.length, timeout_);
+    }
+    /**
+     * Polling on items with given selector
+     * CAUTION: This could be affected by jdk epoll bug
+     *
+     * @param selector Open and reuse this selector and do not forget to close when it is not used.
+     * @param items_
+     * @param count
+     * @param timeout_
+     * @return number of events
+     */
+    public static int zmq_poll(Selector selector, PollItem[] items_, int count, long timeout_)
+    {
+        if (items_ == null) {
+            ZError.errno (ZError.EFAULT);
+            throw new IllegalArgumentException();
+        }
+        if (count == 0) {
+            if (timeout_ == 0)
+                return 0;
+            try {
+                Thread.sleep (timeout_);
+            } catch (InterruptedException e) {
+            }
+            return 0;
+        }
+        long now = 0;
+        long end = 0;
+        
+        HashMap<SelectableChannel, SelectionKey> saved = new HashMap<SelectableChannel, SelectionKey> ();
+        for (SelectionKey key: selector.keys ()) {
+            if (key.isValid ())
+                saved.put(key.channel (), key);
+        }
+
+        for (int i = 0; i < count; i++) {
+            PollItem item = items_[i];
+            if (item == null) 
+                continue;
+
+            SelectableChannel ch = item.getChannel (); // mailbox channel if ZMQ socket
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *ch*. Currently, there are execution paths that do not contain closure statements, for example, when SelectableChannel.register throws an exception, and an else condition at line 650 evaluates to true. Ensure *ch* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).